src/server/editorServices.ts

```diff
@@ -541,174 +576,162 @@ namespace ts.server {
-            });
+        /* @internal  */
+        onTypeRootFileChanged(project: ConfiguredProject, fileName: NormalizedPath) {
+            project.getCachedServerHost().addOrDeleteFileOrFolder(fileName);
+            project.updateTypes();
+            this.delayUpdateProjectGraphAndInferredProjectsRefresh(project);
         }
 
         /**
          * This is the callback function when a watched directory has added or removed source code files.
          * @param project the project that associates with this directory watcher
          * @param fileName the absolute file name that changed in watched directory
          */
-        private onSourceFileInDirectoryChangedForConfiguredProject(project: ConfiguredProject, fileName: string) {
+        /* @internal */
+        onFileAddOrRemoveInWatchedDirectoryOfProject(project: ConfiguredProject, fileName: NormalizedPath) {
+            project.getCachedServerHost().addOrDeleteFileOrFolder(fileName);
+            const configFilename = project.getConfigFilePath();
+
             // If a change was made inside "folder/file", node will trigger the callback twice:
             // one with the fileName being "folder/file", and the other one with "folder".
             // We don't respond to the second one.
             if (fileName && !isSupportedSourceFileName(fileName, project.getCompilerOptions(), this.hostConfiguration.extraFileExtensions)) {
+                this.logger.info(`Project: ${configFilename} Detected file add/remove of non supported extension: ${fileName}`);
                 return;
             }
 
-            this.logger.info(`Detected source file changes: ${fileName}`);
-            this.throttledOperations.schedule(
-                project.getConfigFilePath(),
-                /*delay*/250,
-                () => this.handleChangeInSourceFileForConfiguredProject(project, fileName));
+            const configFileSpecs = project.configFileSpecs;
+            const result = getFileNamesFromConfigSpecs(configFileSpecs, getDirectoryPath(configFilename), project.getCompilerOptions(), project.getCachedServerHost(), this.hostConfiguration.extraFileExtensions);
+            const errors = project.getAllProjectErrors();
+            const isErrorNoInputFiles = (error: Diagnostic) => error.code === Diagnostics.No_inputs_were_found_in_config_file_0_Specified_include_paths_were_1_and_exclude_paths_were_2.code;
+            if (result.fileNames.length !== 0) {
+                filterMutate(errors, error => !isErrorNoInputFiles(error));
+            }
+            else if (!configFileSpecs.filesSpecs && !some(errors, isErrorNoInputFiles)) {
+                errors.push(getErrorForNoInputFiles(configFileSpecs, configFilename));
+            }
+            this.updateNonInferredProjectFiles(project, result.fileNames, fileNamePropertyReader, /*clientFileName*/ undefined);
+            this.delayUpdateProjectGraphAndInferredProjectsRefresh(project);

```

Comment:
Because with add/remove file the project structure could change and open file may or may not be part of the project so the inferred project refresh helps in keeping all the project structures upto date